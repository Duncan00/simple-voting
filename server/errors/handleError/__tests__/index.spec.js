'use strict';

const {ERROR_KEYS, makeError, HTTP_RESPONSES} = require('../../index');
const handleError = require('../');
const _ = require('lodash');

describe('handleError', () => {
	const ctx = {};

	test('success', async () => {
		const next = jest.fn().mockResolvedValue();

		await handleError()(ctx, next);
		expect(next).toBeCalledTimes(1);
	});

	test('meta_code globally unique', () => {
		const codes = Object.values(ERROR_KEYS).map(
			k => HTTP_RESPONSES[k].meta_code
		);
		expect(codes).toEqual(_.uniq(codes));
		expect(codes).toMatchSnapshot('list of meta_codes');
	});

	test('schema validation error', async () => {
		const next = jest.fn(async () => {
			const mock_ajv_err = new Error('Schema validation error');
			mock_ajv_err.isSwaggerAjvValidationError = true;
			mock_ajv_err.details = [
				{
					path: 'data.user_id',
					info: 'data.user_id is a required property'
				},
				{
					path: 'data.started',
					info: 'data.started is a required property'
				}
			];
		});
		await handleError()(ctx, next);
		expect(ctx).toMatchSnapshot('rendered error response');
	});

	describe('handleKnownError', () => {
		Object.values(ERROR_KEYS)
			.filter(k => k !== ERROR_KEYS.INTERNAL)
			.forEach(key => {
				test(`ERROR_KEY = ${key.toString()}`, async () => {
					const err = makeError(key);
					const next = jest.fn().mockRejectedValueOnce(err);
					await handleError()(ctx, next);
					expect(ctx).toMatchSnapshot('> http response of the error');
				});
				test(`render error details for ${key.toString()}`, async () => {
					const err = makeError(key).withDetail(
						null,
						'Additional error message is shown.'
					);
					const next = jest.fn().mockRejectedValueOnce(err);
					await handleError()(ctx, next);
					expect(ctx).toMatchSnapshot('> http response of the error');
				});
			});
	});

	describe('handleUnknownError', () => {
		test('for err.details not an array, hide it in http reponse', async () => {
			ctx.config = {
				env: 'production'
			};
			ctx.app = {
				emit: jest.fn()
			};
			const err = new Error('whatever');
			err.details = 'lorem ipsum error details in text';
			const next = jest.fn().mockImplementation(async () => {
				throw err;
			});

			await handleError()(ctx, next);
			expect(ctx.body).toMatchSnapshot('> http response of the error');
		});
		test('unexpected error to match correct ctx', async () => {
			ctx.config = {
				env: 'development'
			};
			ctx.app = {
				emit: jest.fn()
			};
			const err = new Error('test error');
			const next = jest.fn().mockImplementation(async () => {
				throw err;
			});

			await handleError()(ctx, next);
			expect(ctx.app.emit).toBeCalledWith('error', err, ctx);
			expect(ctx).toEqual(
				Object.assign(ctx, {
					body: {
						data: {},
						meta: {
							code: 500,
							details: err.stack
						}
					},
					status: 500
				})
			);
		});

		test('unexpected error to match correct ctx with different env', async () => {
			ctx.config = {
				env: 'test'
			};
			ctx.app = {
				emit: jest.fn()
			};
			const err = new Error('test error');
			const next = jest.fn().mockImplementation(async () => {
				throw err;
			});

			await handleError()(ctx, next);
			expect(ctx.app.emit).toBeCalledWith('error', err, ctx);
			expect(ctx).toEqual(
				Object.assign(ctx, {
					body: {
						data: {},
						meta: {
							code: 500,
							details: [],
							message: 'Internal Error'
						}
					},
					status: 500
				})
			);
		});
	});
});
