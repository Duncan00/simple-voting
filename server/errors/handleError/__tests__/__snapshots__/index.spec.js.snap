// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleError handleKnownError ERROR_KEY = Symbol(AUTH_ERROR): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 40100,
      "errors": Array [],
      "message": "Authentication Error.",
      "type": "AuthenticationError",
    },
  },
  "status": 401,
}
`;

exports[`handleError handleKnownError ERROR_KEY = Symbol(CANDIDATE_NAMES_CONFLICT): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 40901,
      "errors": Array [],
      "message": "Candidate names are conflict.",
      "type": "CANDIDATE_NAMES_CONFLICT",
    },
  },
  "status": 409,
}
`;

exports[`handleError handleKnownError ERROR_KEY = Symbol(CONFLICT): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 40900,
      "errors": Array [],
      "message": "Conflict.",
      "type": "Conflict",
    },
  },
  "status": 409,
}
`;

exports[`handleError handleKnownError ERROR_KEY = Symbol(NOT_FOUND): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 40400,
      "errors": Array [],
      "message": "Resource is not found.",
      "type": "NotFound",
    },
  },
  "status": 404,
}
`;

exports[`handleError handleKnownError ERROR_KEY = Symbol(SCHEMA_VALIDATION_ERROR): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 42201,
      "errors": Array [],
      "message": "Schema Validation Error",
      "type": "SchemaValidationError",
    },
  },
  "status": 422,
}
`;

exports[`handleError handleKnownError ERROR_KEY = Symbol(UNPROCESSABLE_ENTITY): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 42200,
      "errors": Array [],
      "message": "Unprocessable Entity.",
      "type": "UnprocessableEntity",
    },
  },
  "status": 422,
}
`;

exports[`handleError handleKnownError render error details for Symbol(AUTH_ERROR): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 40100,
      "errors": Array [
        Object {
          "info": "Additional error message is shown.",
          "path": null,
        },
      ],
      "message": "Authentication Error.",
      "type": "AuthenticationError",
    },
  },
  "status": 401,
}
`;

exports[`handleError handleKnownError render error details for Symbol(CANDIDATE_NAMES_CONFLICT): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 40901,
      "errors": Array [
        Object {
          "info": "Additional error message is shown.",
          "path": null,
        },
      ],
      "message": "Candidate names are conflict.",
      "type": "CANDIDATE_NAMES_CONFLICT",
    },
  },
  "status": 409,
}
`;

exports[`handleError handleKnownError render error details for Symbol(CONFLICT): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 40900,
      "errors": Array [
        Object {
          "info": "Additional error message is shown.",
          "path": null,
        },
      ],
      "message": "Conflict.",
      "type": "Conflict",
    },
  },
  "status": 409,
}
`;

exports[`handleError handleKnownError render error details for Symbol(NOT_FOUND): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 40400,
      "errors": Array [
        Object {
          "info": "Additional error message is shown.",
          "path": null,
        },
      ],
      "message": "Resource is not found.",
      "type": "NotFound",
    },
  },
  "status": 404,
}
`;

exports[`handleError handleKnownError render error details for Symbol(SCHEMA_VALIDATION_ERROR): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 42201,
      "errors": Array [
        Object {
          "info": "Additional error message is shown.",
          "path": null,
        },
      ],
      "message": "Schema Validation Error",
      "type": "SchemaValidationError",
    },
  },
  "status": 422,
}
`;

exports[`handleError handleKnownError render error details for Symbol(UNPROCESSABLE_ENTITY): > http response of the error 1`] = `
Object {
  "body": Object {
    "data": Object {},
    "meta": Object {
      "code": 42200,
      "errors": Array [
        Object {
          "info": "Additional error message is shown.",
          "path": null,
        },
      ],
      "message": "Unprocessable Entity.",
      "type": "UnprocessableEntity",
    },
  },
  "status": 422,
}
`;

exports[`handleError handleUnknownError for err.details not an array, hide it in http reponse: > http response of the error 1`] = `
Object {
  "data": Object {},
  "meta": Object {
    "code": 50000,
    "errors": Array [],
    "message": "Internal Error.",
    "type": "InternalError",
  },
}
`;

exports[`handleError meta_code globally unique: list of meta_codes 1`] = `
Array [
  40900,
  50000,
  42201,
  40100,
  40400,
  42200,
  40901,
]
`;

exports[`handleError schema validation error: rendered error response 1`] = `Object {}`;
